crates/ruff_python_formatter/src/expression/expr_tuple.rs

```diff
@@ -1,19 +1,156 @@
+        // Tell the children they need parentheses
+        f.context_mut().set_node_level(NodeLevel::Expression);
+
+        let magic_trailing_comma = USE_MAGIC_TRAILING_COMMA
+            && matches!(
+                find_first_non_trivia_character_after(last.range().end(), f.context().contents()),
+                Some((_, ','))
+            );
+
+        if magic_trailing_comma {
+            // A magic trailing comma forces us to print in expanded mode since we have more than
+            // one element
+            write!(
+                f,
+                [group(&format_args![
+                    // An expanded group always needs parentheses
+                    &text("("),
+                    hard_line_break(),
+                    block_indent(&ExprSequence::new(elts)),
+                    hard_line_break(),
+                    &text(")"),
+                ])]
+            )?;
+        } else if is_parenthesized(*range, elts, f) {
+            // If the tuple has parentheses, keep them. Also only top level tuples are allowed to
+            // elide them
+            write!(
+                f,
+                [group(&format_args![
+                    // If it was previously parenthesized, add them again
+                    &text("("),
+                    soft_block_indent(&ExprSequence::new(elts)),
+                    &text(")"),
+                ])]
+            )?;
+        } else {
+            write!(
+                f,
+                [group(&format_args![
+                    // If there were previously no parentheses, add them only if the group breaks
+                    if_group_breaks(&text("(")),
+                    soft_block_indent(&ExprSequence::new(elts)),
+                    if_group_breaks(&text(")")),
+                ])]
+            )?;
+        }
+
+        f.context_mut().set_node_level(saved_level);
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct ExprSequence<'a> {
+    elts: &'a [Expr],
+}
+
+impl<'a> ExprSequence<'a> {
+    const fn new(elts: &'a [Expr]) -> Self {
+        Self { elts }
+    }
+}
+
+impl Format<PyFormatContext<'_>> for ExprSequence<'_> {
+    fn fmt(&self, f: &mut Formatter<PyFormatContext<'_>>) -> FormatResult<()> {
+        for (pos, entry) in self.elts.iter().enumerate() {
+            // We need a trailing comma on the last entry of an expanded group since we have more
+            // than one element
+            if pos == self.elts.len() - 1 {
+                write!(
+                    f,
+                    [
+                        entry.format(),
+                        if_group_breaks(&text(",")),
+                        soft_line_break()
+                    ]
+                )?;
+            } else {
+                write!(f, [entry.format(), text(","), soft_line_break_or_space()])?;
+            }
+        }
+        Ok(())
     }
 }
 
+/// Check if a tuple has already had parentheses in the input
+fn is_parenthesized(
+    range: TextRange,
```

Comment:
```suggestion
    tuple_range: TextRange,
```

I had to look it up on the call side to know what range this is.It could even make sense to pass the whole tuple